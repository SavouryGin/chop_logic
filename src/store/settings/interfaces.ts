import { Language } from 'types';

export interface SettingsInitialState {
  language: Language;
  flags: SettingsFlags;
}

export interface SettingsFlags {
  isNavigationOpen: boolean;
  isSidebarOpen: boolean;
  isDarkMode: boolean;
  isFullScreen: boolean;
  isSettingOpen: boolean;
  isSoundsEnabled: boolean;
  isMenuAnimationActive: boolean;
  isSidebarAnimationActive: boolean;
  isModalWindowClosingAnimationActive: boolean;
}

export type SettingsFlag = keyof SettingsFlags;
