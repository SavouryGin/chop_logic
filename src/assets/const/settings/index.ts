import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
};
