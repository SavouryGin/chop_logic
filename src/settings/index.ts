import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const languageStringOptions = ['en', 'ru'];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  linkedIn: 'https://www.linkedin.com/in/dmitrii-suroviagin/',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
  facebook: 'https://www.facebook.com/dmitrii.suroviagin',
};
