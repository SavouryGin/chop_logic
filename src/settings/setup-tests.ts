import crypto from 'crypto';
import '@testing-library/jest-dom/extend-expect';

Object.defineProperty(globalThis, 'crypto', {
  value: {
    randomUUID: () => crypto.randomUUID(),
  },
});
