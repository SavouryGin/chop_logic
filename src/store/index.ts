import createSagaMiddleware from 'redux-saga';
import rootReducer from './root-reducer';
import rootSaga from './root-saga';
import { configureStore } from '@reduxjs/toolkit';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([sagaMiddleware]),
});

sagaMiddleware.run(rootSaga);
