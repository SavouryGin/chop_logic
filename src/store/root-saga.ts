import propositionsDPSagas from './propositions/direct-proofs/sagas';
import propositionsNPSagas from './propositions/natural-proofs/sagas';
import truthTableSagas from './propositions/truth-tables/sagas';
import { all } from 'redux-saga/effects';

function* rootSaga(): Generator {
  yield all([propositionsDPSagas(), propositionsNPSagas(), truthTableSagas()]);
}

export default rootSaga;
