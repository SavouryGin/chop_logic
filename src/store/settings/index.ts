import { Language } from 'types';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { SettingsFlag, SettingsInitialState } from './interfaces';

export const settingsInitialState: SettingsInitialState = {
  language: 'en',
  flags: {
    isNavigationOpen: false,
    isSidebarOpen: false,
    isDarkMode: false,
    isFullScreen: false,
    isSettingOpen: false,
    isSoundsEnabled: false,
  },
};

export const settingsSlice = createSlice({
  name: 'settings',
  initialState: settingsInitialState,
  reducers: {
    resetSettings: () => {
      return settingsInitialState;
    },

    toggleFlag: (state, action: PayloadAction<SettingsFlag>) => {
      const currentFlagValue = state.flags[action.payload];
      state.flags[action.payload] = !currentFlagValue;
    },

    setLanguage: (state, action: PayloadAction<Language>) => {
      state.language = action.payload;
    },

    setFullScreenFlag: (state, action: PayloadAction<boolean>) => {
      state.flags.isFullScreen = action.payload;
    },

    setUpFlag: (state, action: PayloadAction<{ flag: SettingsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },
  },
});

export const settingsActions = { ...settingsSlice.actions };
