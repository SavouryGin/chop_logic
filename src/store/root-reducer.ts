import { combineReducers } from '@reduxjs/toolkit';
import { propositionsDPSlice } from './propositions/direct-proofs';
import { propositionsNPSlice } from './propositions/natural-proofs';
import { settingsSlice } from './settings/slice';

const rootReducer = combineReducers({
  settings: settingsSlice.reducer,
  propositionsDP: propositionsDPSlice.reducer,
  propositionsNP: propositionsNPSlice.reducer,
});

export default rootReducer;
