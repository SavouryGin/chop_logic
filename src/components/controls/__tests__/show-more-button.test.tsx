import React from 'react';
import ShowMoreButton from '../show-more-button';
import renderWithRedux from 'utils/test-utils/render-with-redux';
import userEvent from '@testing-library/user-event';
import { combineReducers } from '@reduxjs/toolkit';
import { screen } from '@testing-library/react';
import { settingsInitialState, settingsSlice } from 'store/settings';

describe('ShowMoreButton component:', () => {
  const testProps = {
    onClick: jest.fn(),
    isOpened: true,
  };

  const mockedReducer = combineReducers({ settings: settingsSlice.reducer });
  const mockedState = { settings: settingsInitialState };

  it('renders the button element with the default type', () => {
    renderWithRedux(<ShowMoreButton {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveProperty('type', 'button');
  });

  it('has the passed class names', () => {
    renderWithRedux(
      <ShowMoreButton {...testProps} className={{ 'test-class-1': true, 'test-class-2': false, 'test-class-3': true }} />,
      mockedReducer,
      mockedState,
    );
    expect(screen.getByRole('button')).toHaveClass('show-more-button', 'test-class-1', 'test-class-3', 'chop-icon__caret-up');
  });

  it('fires the onClick handler', async () => {
    renderWithRedux(<ShowMoreButton {...testProps} />, mockedReducer, mockedState);
    const btn = screen.getByRole('button');
    await userEvent.click(btn);
    expect(testProps.onClick).toHaveBeenCalledTimes(1);
    await userEvent.dblClick(btn);
    expect(testProps.onClick).toHaveBeenCalledTimes(3);
  });

  it('could be disabled', () => {
    renderWithRedux(<ShowMoreButton {...testProps} isDisabled={true} />, mockedReducer, mockedState);
    expect(screen.getByRole('button')).toBeDisabled();
  });

  it('should match the snapshot', () => {
    const { asFragment } = renderWithRedux(<ShowMoreButton {...testProps} />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });
});
