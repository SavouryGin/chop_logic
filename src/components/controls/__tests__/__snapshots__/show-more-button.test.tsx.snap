// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMoreButton component: should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="show-more-button chop-icon__caret-up"
    data-testid="show-more-button"
    type="button"
  />
</DocumentFragment>
`;
